Best practices for spaces (#695)
OctopusDeploy/blog: blog/2022-02/best-practices-spaces/index.md

---

> title: Best practices for Spaces

Reviewer: I wonder if this is slightly more descriptive
```suggestion
title: Best practices for Octopus Spaces
```

description: Find out how to use Spaces effectively.
author: steve.fenton@octopus.com
visibility: private
published: 9999-01-01
metaImage:  blogimage-spaces.png
bannerImage: blogimage-spaces.png
tags:
 - Product
 - Spaces
---

> The Octopus Spaces feature helps you organize and secure your projects, environments, and infrastructure and control which team members can access them. They are a useful alternative to installing multiple Octopus Deploy instances.

Reviewer: I really like the opening you have here, Steve. Clear, concise and direct, but I might like to add a line or two first that speaks directly to the use case. For instance:

```suggestion
Over the years at Octopus, we've always been pleased to see our customers expand their use of Octopus Deploy to include new projects and teams. In the early days, it was hard to imagine this could ever pose any kind of problem. However, there comes a point when having too many projects in Octopus can become a distraction, particularly for a team that only works on a handful of those projects.

The Octopus Spaces feature helps you organize and secure your projects, environments, and infrastructure and control which team members can access them. They are a useful alternative to installing multiple Octopus Deploy instances.
```

Author: This is great because it actually starts with the problem. It sounds less robotic than the original!

Spaces give you a convenient way to control access to a group of related resources without complicated permissions. You can grant a team member full or read-only access to the spaces they need, and they can then switch between them quickly using the space switcher.

![The space switcher appears in the top-left of the Octopus Deploy navigation bar](space-switcher.jpg)

## How to design your spaces

A space represents a logical group of applications. If you have several closely related components, they are likely to share some variables in a variable set or be deployed to the same infrastructure. These factors will naturally guide whether they should be grouped into a space or kept in separate spaces to prevent them from becoming related in undesirable ways.

It may seem like a good idea to create a space for each team in your organization, but this is not always the best design to follow. In particular, this may lead you to split items into too many spaces in situations where multiple teams contribute to the same application.

Instead, use one of the following dimensions to design your spaces:

- Clients
- Application groups
- Application audiences
- Company divisions

> More information on each of these options is below.

Reviewer: ```suggestion
Read on for more information about each of these options.
```

:::hint
You can use the [export/import project feature](https://octopus.com/docs/projects/export-import) to move projects to a new space.
:::

### Client

> If you are an agency managing deployments for multiple clients, using a space per client ensures no data will be shared between customers. All deployment targets, projects, variables, and lifecycles will be exclusive to the customer space.

Reviewer: Ah I see you do mention them further down :D

You may decide to sub-divide a client into multiple spaces using one of the other options.

### Application groups

An application group is a set of related components that are likely to be deployed to the same location.

Application groups are ideal for organizing spaces because similar design considerations apply to spaces that will have influenced your application design. For example, you might group the components of a content management system (CMS) within one space and the components of a billing system into a second space.

You may find that this design for spaces matches how you have organized your teams. It is fine to land on this solution if you have done so by considering your application groups.

> If you have more than one team contributing to an application group, you should maintain a space that aligns with the software, not the teams. You can give both teams access to the space, and both can see a whole-system view of deployments.

Author: Note to self - I think I might fix "both teams" as it assumes two, it could be more. I'll leave this open to remind me.

### Application audiences

> A less granular approach is to have separate spaces that match intended audiences, for example, internal and public-facing applications. This approach will clean up the public-facing space by moving internal resources into a different one.

Reviewer: I feel the repetition pain, but in this instance, I think it makes the text clearer :)

```suggestion
A less granular approach is to have separate spaces that match intended audiences, for example, internal and public-facing applications. This approach will clean up the public-facing space by moving internal resources into a different space.
```

With their own space manager, each division would be self-sufficient in managing its space without cluttering Octopus Deploy for other divisions.

## Useful design indicators

The ideal scenario is that applications within a space are independent, deployed to dedicated targets, and have an autonomous team responsible for them. While you might not find yourself in this perfect situation, it provides a helpful guide when deciding how to design your spaces.

If you deploy multiple applications to the same deployment targets, you should keep the deployments within the same space. It is possible to set up the same deployment target in more than one space using a listening tentacle. However, this complicates the permissions, and team members won't see all the deployments targeting the shared infrastructure.

If there are strong reasons to split the deployments into multiple spaces, those same reasons are likely to mean you should have different deployment targets, too.

## What to avoid

> You should avoid using spaces for each environment, as you will need to duplicate the process in each space. It would not be easy to keep the process consistent in each space when you make changes and you won't benefit from release snapshots. Release snapshots make sure the same version of the package, variables, and process is used throughout environments, and this improves the reliability of your deployments.

Reviewer: Perhaps a personal preference for stating what things are, as opposed to what they're not :) 

```suggestion
You should avoid using spaces for each environment, as you will need to duplicate the process in each space. It would be difficult to keep the process consistent in each space when you make changes and you won't benefit from release snapshots. Release snapshots make sure the same version of the package, variables, and process is used throughout environments, and this improves the reliability of your deployments.
```

> You should also avoid using spaces where it is more appropriate to use [tenants](https://octopus.com/docs/tenants).

Reviewer: Contractions give us a more casual/friendly tone:
```suggestion
You should also avoid using spaces where it's more appropriate to use [tenants](https://octopus.com/docs/tenants).
```
